keys.go

```diff
@@ -175,40 +282,190 @@ func (k pkcs11PrivateKey) Public() crypto.PublicKey {
+
+	if _, ok := attributes[CkaClass]; ok {
+		return nil, errors.Errorf("key attribute set must not contain CkaClass")
+	}
 
 	err := c.withSession(func(session *pkcs11Session) error {
-		privHandle, err := findKey(session, id, label, uintPtr(pkcs11.CKO_SECRET_KEY), nil)
+		// Add the private key class to the template to find the private half
+		privAttributes := attributes.Copy()
+		privAttributes[CkaClass] = pkcs11.NewAttribute(CkaClass, pkcs11.CKO_SECRET_KEY)
+
+		privHandles, err := findKeysWithAttributes(session, privAttributes.ToSlice())
 		if err != nil {
 			return err
 		}
-		if privHandle == nil {
-			// Key does not exist
-			return nil
+
+		for _, privHandle := range(privHandles) {
+			attributes := []*pkcs11.Attribute{
+				pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, 0),
+			}
+			if attributes, err = session.ctx.GetAttributeValue(session.handle, privHandle, attributes); err != nil {
+				return err
+			}
+			keyType := bytesToUlong(attributes[0].Value)
+
+			if cipher, ok := Ciphers[int(keyType)]; ok {
+				k := &SecretKey{pkcs11Object{privHandle, c}, cipher}
+				keys = append(keys, k)
+			} else {
+				return errors.Errorf("unsupported key type: %X", keyType)
+			}
 		}
 
-		attributes := []*pkcs11.Attribute{
-			pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, 0),
+		return nil
+	})
+
+	return keys, err
+}
+
+// FindAllKeyPairs retrieves all existing symmetric keys, or a nil slice if none can be found.
+func (c *Context) FindAllKeys() ([]*SecretKey, error) {
+	return c.FindKeysWithAttributes(NewAttributeSet())
+}
+
+func uintPtr(i uint) *uint { return &i }
+
+func (c *Context) getAttributes(handle pkcs11.ObjectHandle, attributes []AttributeType) (a AttributeSet, err error) {
+	values := NewAttributeSet()
+
+	err = c.withSession(func(session *pkcs11Session) error {
+		var attrs []*pkcs11.Attribute
+		for _, a := range(attributes) {
+			attrs = append(attrs, pkcs11.NewAttribute(a, nil))
 		}
-		if attributes, err = session.ctx.GetAttributeValue(session.handle, *privHandle, attributes); err != nil {
+
+		p11values, err := session.ctx.GetAttributeValue(session.handle, handle, attrs)
+		if err != nil {
 			return err
 		}
-		keyType := bytesToUlong(attributes[0].Value)
 
-		if cipher, ok := Ciphers[int(keyType)]; ok {
-			k = &SecretKey{pkcs11Object{*privHandle, c}, cipher}
-		} else {
-			return errors.Errorf("unsupported key type: %X", keyType)
+		for _, a := range(p11values) {
+			values[a.Type] = pkcs11.NewAttribute(a.Type, a.Value)
 		}
+
 		return nil
 	})
 
-	return k, err
+	return values, err
 }
 
-func uintPtr(i uint) *uint { return &i }
+// GetAttributes gets the values of the specified attributes on the given key
+// If the key is asymmetric, then the attributes are retrieved from the private half
+func (c *Context) GetAttributes(key interface{}, attributes []AttributeType) (a AttributeSet, err error) {
+	if c.closed.Get() {
+		return nil, errClosed
+	}
+
+	var handle pkcs11.ObjectHandle
+
+	switch k := (key).(type) {
```

Comment:
I tried something like this first before the switch, but it doesn't seem to work, because while the object contains a `pkcs11Object` it is not itself a `pkcs11Object` so the type assertion fails.